Rails 5.2 changes (many needed for 6.1 support)

kbrock opened the pull request:
Lots of comment changes.
Some deprecations of rails 5.1 behavior

A few quick hits that are needed for rails 6.1 support.

commits: 

- stop using join_keys 
- comments to understand what changed in the cod
- more drop rails 5.1 changes
- drop deprecation warnings when associations are invalid 
- allow more recent versions of db query matchers (dev only)
- remove unneeded Arel::Attribute.to_sql 
- alphabetize gemspec
- remove send(:include)

kbrock commented:
re Hakiri: I am not going to limit rails versions here. that is for the app not a gem to decide

jrafanie commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -363,7 +328,9 @@ def apply_join_dependency(*args, **kargs, &block)
       # From ActiveRecord::QueryMethods (rails 5.2 - 6.0)
       def build_select(arel)
         if select_values.any?
+          # change (only place allow_alias is true - the select clause)
```
I'm not sure what this means: `(only place allow_alias is true - the select clause)`... maybe say why there is a change here?  I can't figure it out at a glance.

jrafanie commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -376,11 +343,15 @@ def arel_columns(columns, allow_alias = false)
         columns.flat_map do |field|
           case field
           when Symbol
+            # change
```
Again, maybe just say why there is a change for future you/me/others.

jrafanie commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -376,11 +343,15 @@ def arel_columns(columns, allow_alias = false)
               connection.quote_table_name(attr_name)
             end
           when String
+            # change
```
Ditto.  Why?

jrafanie commented on lib/active_record/virtual_attributes/virtual_total.rb:
```diff
@@ -151,4 +151,4 @@ def virtual_aggregate_arel(reflection, method_name, column)
 end
 
-ActiveRecord::Base.send(:include, VirtualAttributes::VirtualTotal)
+ActiveRecord::Base.include VirtualAttributes::VirtualTotal
```
Weird, why did we use send before?

kbrock replied:
A pervious version of ruby made `include` private. Sorry, can't remember the exact version.
Now that we are only supporting more recent versions of ruby, we can use the standard `include`

jrafanie replied:
Ah, ok, thanks!

jrafanie commented on spec/virtual_includes_spec.rb:
```diff
@@ -200,6 +200,22 @@
+    #
+    # wasted work and it is missing author_id which is needed for the join
+    #
+    # think this test is here because we do this bad code miq
```
👍

jrafanie commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -190,10 +190,8 @@ def grouped_records(orig_association, records, polymorphic_parent)
                 reflection = record.class._reflect_on_association(association)
-                display_virtual_attribute_deprecation("#{record.class.name}.#{association} does not exist") if !reflection && !polymorphic_parent
-                next if !reflection || !record.association(association).klass
+                next if polymorphic_parent && !reflection || !record.association(association).klass
```
Interesting.  So, now, we'd do `next` for a nil/false `polymorphic_parent`.  Previously, was this a bug?

jrafanie replied:
This was answered below in https://github.com/ManageIQ/activerecord-virtual_attributes/pull/97#issuecomment-1010195099

kbrock commented:
### `polymorphic_parent` feature

#### Rails 6.0.x

Rails 6.0.x introduced the feature to blow up when `includes()` was passed an invalid value. Our team was frustrated with a few bugs in our code that were due to not properly checking the includes values.

#### Adding Rails 5.2 and 6.0 support

When I introduced support for rails 5.2 and 6.0 in #62 I used the 6.0.x implementation of `group_records` for both versions of rails. It was close enough and I did not have the energy to construct 2 monkeypatchs for two almost identical implementations of this complicated method. Besides, we wanted the protection of the 6.0 implementation.

#### It sounded like a good idea

Well, it turned out that we had more invalid `includes()` than we had originally anticipated, and we were not ready for this change. So we introduced the deprecation warning instead of blowing up in #67 

#### Today
Since over the past 2 years we have fixed any invalid `includes()`, this PR removes the deprecation warning and goes back to standard rails 6.0. Well, standard if you ignore the fact that we added the virtual includes.

kbrock commented:
push: added comments, cops, and fixed merge conflict

jrafanie commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -372,15 +340,21 @@ def build_select(arel)
 
       # from ActiveRecord::QueryMethods (rails 5.2 - 6.0)
+      # @param allow_alias include AS alias for the generated SQL
+      #        valis is true for SELECT and false for all others (e.g.: GROUP BY or COUNT)
```
typo:  valis == value?

jrafanie commented on lib/active_record/virtual_attributes/virtual_fields.rb:
```diff
@@ -390,19 +364,24 @@ def arel_columns(columns, allow_alias = false)
 
       # from ActiveRecord::QueryMethods (rails 5.2 - 6.0)
+      # @param allow_alias include AS alias for the generated SQL
+      #        valis is true for SELECT and false for all others (e.g.: GROUP BY or COUNT)
```
Same... valis...

jrafanie commented:
> sorry didn't answer:
>> Previously, was this a bug?

I literally 🤣  when I got to this part.  Thanks for noticing you gave a useful history lesson and remembered you still didn't answer that.

> before 62, we were too eager to call next. Well, rails was too eager too
> before 67, we properly were calling the next - thanks to the rails 6.0 fix
> after 67, we rolled back the fix a little.
> This PR reverts the rollback, so we end up with logic closer to what resides in 6.0

😎 Ok, thanks!
